File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/ui/CustomModifiersArePrivate.kt
package com.perrystreet.woof.konsist.ui

import com.lemonappdev.konsist.api.ext.list.modifierprovider.withPublicOrDefaultModifier
import com.perrystreet.woof.konsist.Assertions.assertEmpty
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class CustomModifiersArePrivate : BehaviorSpec() {
    init {
        Given("A custom Modifier extension") {
            val modifiers = KonsistUtils.productionCode.functions(includeNested = true)
                .filter { it.receiverType?.name == "Modifier" }
                .withPublicOrDefaultModifier()
                .filter { it.name !in Baseline }

            Then("It is private") {
                modifiers.assertEmpty(message = Message)
            }
        }
    }

    private companion object {
        private val Baseline = setOf(
            "sectionContainer",
        )

        private val Message = LintRuleMessage(
            rule = "Custom Modifier extensions are private.",
            why = "A public modifier spreads styling outside the component that owns it, so callers restyle one-off instead of picking a variant.",
            howToFix = "Make the modifier private to its component. If several components need it, move them into that file as overloads of one component.",
            badExample = """
                fun Modifier.atomPlaceholderShimmer(): Modifier = ...
            """.trimIndent(),
            goodExample = """
                @Composable
                fun AtomPlaceholder(role: TextPlaceholderRole) {
                    Box(modifier = Modifier.placeholderShimmer(role.radius()))
                }

                @Composable
                fun AtomPlaceholder(role: ShapePlaceholderRole) {
                    Box(modifier = Modifier.placeholderShimmer(role.radius()))
                }

                private fun Modifier.placeholderShimmer(radius: Dp): Modifier = ...
            """.trimIndent(),
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/UI/CustomModifiersArePrivate.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class CustomModifiersArePrivate: QuickSpec {
    override class func spec() {
        Given("A custom view modifier") {
            let modifiers = WoofHarmonize.productionCode.structs()
                .withPublicModifier()
                .filter { $0.inheritanceTypesNames.contains("ViewModifier") && !baseline.contains($0.name) }

            Then("It is not public") {
                modifiers.assertEmpty(message: message)
            }
        }

        Given("A View extension function") {
            let functions = WoofHarmonize.productionCode.extensions()
                .filter { $0.typeAnnotation?.name == "View" }
                .flatMap { viewExtension in
                    viewExtension.functions.filter { viewExtension.modifiers.contains(.public) || $0.modifiers.contains(.public) }
                }
                .filter { !baseline.contains($0.name) }

            Then("It is not public") {
                functions.assertEmpty(message: message)
            }
        }
    }

    private static let baseline: Set<String> = [
        "OrgAlertDialog",
        "orgAlertDialog",
        "ErrorAdapter",
        "errorAdapter",
    ]

    private static let message = LintRuleMessage(
        rule: "Custom view modifiers are private.",
        why: "A public modifier spreads styling outside the component that owns it, so callers restyle one-off instead of picking a variant.",
        howToFix: "Make the modifier private to its component. If several components need it, move them into that file as initializers of one component.",
        badExample: """
        public extension View {
            func atomPlaceholderShimmer() -> some View { modifier(AtomPlaceholderShimmer()) }
        }
        """,
        goodExample: """
        public struct AtomPlaceholder: View {
            public init(role: TextPlaceholderRole) { ... }
            public init(role: ShapePlaceholderRole) { ... }
        }

        private extension View {
            func placeholderShimmer(radius: CGFloat) -> some View { modifier(PlaceholderShimmer(radius: radius)) }
        }
        """
    )
}
